daemon/cluster/listen_addr.go

```diff
@@ -149,6 +151,32 @@ func resolveInterfaceAddr(specifiedInterface string) (net.IP, error) {
 	return interfaceAddr6, nil
 }
 
+// resolveInputIPAddr tries to resolve the IP address from the string passed as input
+// - tries to match the string as an interface name, if so returns the IP address associated with it
+// - on failure of previous step tries to parse the string as an IP address itself
+//	 if succeeds returns the IP address
+func resolveInputIPAddr(input string, isUnspecifiedValid bool) (net.IP, error) {
+	// Try to see if it is an interface name
+	interfaceAddr, err := resolveInterfaceAddr(input)
+	if err == nil {
+		return interfaceAddr, nil
+	}
+	// String matched interface but there is a potential ambiguity to be resolved
+	if err != errNoSuchInterface {
+		return nil, err
+	}
+
+	// String is not an interface check if it is a valid IP
+	if ip := net.ParseIP(input); ip != nil {
+		if isUnspecifiedValid || !ip.IsUnspecified() {
```

Comment:
These conditions could be folded into the outer `if`.

```go
if ip := net.ParseIP(input); ip != nil && (isUnspecifiedValid || !ip.IsUnspecified()) {
        return ip, nil
}
```

Or perhaps something like this:

```go
ip := net.ParseIP(input)
if ip == nil {
        return nil, errBadNetworkIdentifier
}
if ip.IsUnspecified()  && !isUnspecifiedValid{
        return nil, errBadNetworkIdentifier
}
return return ip, nil
```